geospaas_rest_api/serializers.py

```diff
@@ -1,106 +1,191 @@
-
-
-class GeographicLocationSerializer(serializers.ModelSerializer):
-    """
-    Serializer for GeographicLocation objects
-    """
+import json
+
+import geospaas.catalog.models
+import geospaas.vocabularies.models
+import rest_framework.serializers
+from rest_framework.exceptions import ValidationError
+
+try:
+    import geospaas_processing.tasks as tasks
+except ImportError:
+    tasks = None
+
+
+class TaskResultSerializer(rest_framework.serializers.Serializer):
+    """Serializer for TaskResult objects"""
+
+    # Fields from the TaskResult model; used to display tasks
+    task_id = rest_framework.serializers.CharField(read_only=True)
+    task_name = rest_framework.serializers.CharField(read_only=True)
+    result = rest_framework.serializers.CharField(read_only=True)
+    status = rest_framework.serializers.CharField(read_only=True)
+    task_args = rest_framework.serializers.CharField(read_only=True)
+    date_created = rest_framework.serializers.DateTimeField(read_only=True)
+    date_done = rest_framework.serializers.DateTimeField(read_only=True)
+    meta = rest_framework.serializers.CharField(read_only=True)
+
+    # Fields used to launch tasks
+    action = rest_framework.serializers.ChoiceField(choices=['download', 'convert'],
+                                                    required=True, write_only=True,
+                                                    help_text="Action to perform")
+    parameters = rest_framework.serializers.DictField(write_only=True,
+                                                      help_text="Parameters for the action")
+
+    def __init__(self, *args, **kwargs):
+        """
+        Check for geospaas_processing.tasks availability
+        before allowing to instantiate the serializer
+        """
+        if not tasks:
+            raise ImportError('geospaas_processing.tasks is not available')
+        super().__init__(*args, **kwargs)
+
+    def to_representation(self, instance):
+        """Adapt representation of result and meta fields"""
+        representation = super().to_representation(instance)
+        representation['result'] = json.loads(representation['result'])
+        representation['meta'] = json.loads(representation['meta'])
+        return representation
+
+    def update(self, instance, validated_data):
+        """Does nothing. Update of already created tasks is only done by Celery"""
+
+    def create(self, validated_data):
+        """Launches a long-running task, and returns the corresponding AsyncResult"""
+        if validated_data['action'] == 'download':
+            result = tasks.download.delay(validated_data['parameters']['dataset_id'])
+        elif validated_data['action'] == 'convert':
+            if validated_data['parameters']['format'].lower() == 'idf':
+                result = tasks.download.apply_async(
+                    (validated_data['parameters']['dataset_id'],),
+                    link=tasks.convert_to_idf.s())
```

Comment:
That looks a bit too hardcoded. Can that be generalised, for example, to: 
```python
result = tasks.tasks[validated_data['action']](**validated_data['parameters'])
```
So that tasks.tasks is a dictionary:
```python
# in tasks.py
tasks = dict(
    download = tasks.download.delay,
    convert = tasks.download.check_format_and_apply_async
)
def check_format_and_apply_async(dataset_id, format):
            if format.lower() == 'idf':
                result = tasks.download.apply_async((dataset_id,),
                    link=tasks.convert_to_idf.s())
```
Not exactly like that, of course, but to have a mapping between task API name and a function, rather than selecting with if/else in `TaskResultSerializer.create`.

Reply:
I am not entirely convinced by the mapping approach: it seems unnecessarily complicated and harder to read, but I will try it out (I am working on adding an extra step to send the results to the FTP server, so I'll see how it integrates with it).

However, even assuming that we adopt it, I don't think it should be moved to `geospaas_processing.tasks`.
The coupling between the API and `geospaas_processing.tasks` would be less loose: the tasks provided by `geospaas_processing.tasks` are "building blocks" that the API uses. Knowing how these blocks are used is the responsibility of the API.